streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java

```diff
@@ -564,32 +587,59 @@ public static void waitForCompletion(final KafkaStreams streams,
+                                                                           final long waitTime,
+                                                                           final boolean withTimestamp) throws InterruptedException {
+        final List<T> accumData = new ArrayList<>();
         try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
             final TestCondition valuesRead = () -> {
-                final List<KeyValue<K, V>> readData =
-                        readKeyValues(topic, consumer, waitTime, expectedRecords.size());
+                final List<T> readData;
+                if (withTimestamp) {
+                    readData = (List<T>) readKeyValuesWithTimestamp(topic, consumer, waitTime, expectedRecords.size());
+                } else {
+                    readData = (List<T>) readKeyValues(topic, consumer, waitTime, expectedRecords.size());
+                }
                 accumData.addAll(readData);
 
                 // filter out all intermediate records we don't want
-                final List<KeyValue<K, V>> accumulatedActual = accumData.stream().filter(expectedRecords::contains).collect(Collectors.toList());
+                final List<T> accumulatedActual = accumData.stream().filter(expectedRecords::contains).collect(Collectors.toList());
 
                 // still need to check that for each key, the ordering is expected
-                final Map<K, List<KeyValue<K, V>>> finalAccumData = new HashMap<>();
-                for (final KeyValue<K, V> kv : accumulatedActual) {
-                    finalAccumData.computeIfAbsent(kv.key, key -> new ArrayList<>()).add(kv);
+                final Map<K, List<T>> finalAccumData = new HashMap<>();
+                for (final T kv : accumulatedActual) {
+                    finalAccumData.computeIfAbsent(
+                        (K) (withTimestamp ? ((KeyValueTimestamp) kv).key() : ((KeyValue) kv).key),
+                        key -> new ArrayList<>()).add(kv);
                 }
-                final Map<K, List<KeyValue<K, V>>> finalExpected = new HashMap<>();
-                for (final KeyValue<K, V> kv : expectedRecords) {
-                    finalExpected.computeIfAbsent(kv.key, key -> new ArrayList<>()).add(kv);
+                final Map<K, List<T>> finalExpected = new HashMap<>();
+                for (final T kv : expectedRecords) {
+                    finalExpected.computeIfAbsent(
+                        (K) (withTimestamp ? ((KeyValueTimestamp) kv).key() : ((KeyValue) kv).key),
+                        key -> new ArrayList<>()).add(kv);
                 }
 
                 // returns true only if the remaining records in both lists are the same and in the same order
                 // and the last record received matches the last expected record
+                System.err.println(finalAccumData);
+                System.err.println(finalExpected);
```

Comment:
just checking... was this for your debugging, or do we really want to print to stdout during the tests?

Reply:
Good catch!